src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs

```diff
@@ -196,13 +220,17 @@ private async Task ProcessQueueAsync()
                         // will be sent back to the client but they can also be captured via HandleNonMutatingRequestError,
                         // though these errors don't put us into a bad state as far as the rest of the queue goes.
                         // Furthermore we use Task.Run here to protect ourselves against synchronous execution of work
-                        // blocking the request queue for longer periods of time (it enforces parallelizabilty).
-                        _ = WrapStartRequestTaskAsync(Task.Run(() => work.StartRequestAsync(context, cancellationToken), cancellationToken), rethrowExceptions: false);
+                        // blocking the request queue for longer periods of time (it enforces parallelizability).
+                        var pendingTask = WrapStartRequestTaskAsync(Task.Run(() => work.StartRequestAsync(context, cancellationToken), cancellationToken), rethrowExceptions: false);
+
+                        pendingTasks.Add(pendingTask, cancellationTokenSource);
+
+                        _ = pendingTask.ContinueWith(t => pendingTasks.Remove(t), TaskScheduler.Default);
```

Comment:
this is not threadsafe.  teh Add and the Remove can happen on entirely different threads.  this can easily lead to corruption of hte dictionary.  If we do want to go this pattern (which isn't something i'm convinced of yet :)), then both sides need to lock the dictionary to make sure this is safe.

Reply:
Good catch, I'll add a lock

Reply:
Switched to ConcurrentDictionary